### PR TITLE
Fix reading indent style and indent size from VS settings

### PR DESCRIPTION
**Pull request by tmat**
Turns out ISettingsManager stores indent size as Int64 which wasn't converted to Int32.
We were also missing storage location for Indent Style.

Fixes https://github.com/dotnet/roslyn/issues/65375 (https://developercommunity.visualstudio.com/t/Quick-actions-use-wrong-indent/10197217)

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1674009
Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1675408

Fixes https://github.com/dotnet/roslyn/issues/65360

### CODE REVIEW

**Comment by tmat** [comment]
@CyrusNajmabadi @jasonmalinowski @arkalyanms

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by jasonmalinowski — APPROVED** [review]
Thanks for the explanation in the PR description!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs:83]: I don't understand how roaming storage location is useful here?
This doesn't seem to be exposed in Tools → Options, and it's not in AutomationObject.

So its value will always remain `DefaultIndentStyle` and it is never fetched/persisted through import/export.

Anything I'm missing? I'm curious to know the the user-visible effect of this change if you noticed any.

**Review by Youssef1313 — COMMENTED** [review]
Does this PR fix https://github.com/dotnet/roslyn/issues/65360 ? If so, can you modify the description to include "Fixes #65360" so the issue gets auto closed?

File: src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs
Comment (Youssef1313) [inline src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs:123]: Should we instead update the following line to match the type stored in `_settingsManager`? https://github.com/dotnet/roslyn/blob/f814f00d41352cac7ed88d745584191e6ce2323e/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs#L55

**Review by Youssef1313 — COMMENTED** [review]
Also I think we could adjust the following:

https://github.com/dotnet/roslyn/blob/f814f00d41352cac7ed88d745584191e6ce2323e/src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs#L205-L210

We can do it like this:

```csharp
try
{
    value = Convert.ChangeType(value, optionKey.Option.Type);
}
catch
{
    value = null; 
    return false; 
}
```

This basically says if we got different types, try to convert it. Can we convert? Then everything is good. Failed to convert? Nothing we can do to read the option.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs:83]: The option is read from global options here: https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Features/Options/IndentationOptionsStorage.cs,21

Global options end up calling `VisualStudioSettingsOptionPersister` to read the option from VS settings, but that only happens if it has the storage.

File: src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs
Comment (tmat) [reply src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs:123]: I don't think so. The fact that the option is stored as In64 in VS setting is a VS-specific quirk. Quite odd, since there is no reason for it to be 64 bit.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs
Comment (arkalyanms) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs:83]: image

**Review by arkalyanms — APPROVED** [review]

**Comment by arkalyanms** [comment]
@tmat Lets park it here until we get soft approval from QB. Please fill in the ask mode template at https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1675408

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs
Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/FormattingOptions2.cs:83]: Thanks @tmat @arkalyanms for clarifying.

File: src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs
Comment (Youssef1313) [reply src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs:123]: @tmat Then what about the more generalized approach suggested in https://github.com/dotnet/roslyn/pull/65382#pullrequestreview-1178115501?

This will ensure that no such issues exist whatever the type of option is.

Comment (tmat) [reply src/VisualStudio/Core/Def/Options/VisualStudioSettingsOptionPersister.cs:123]: Will follow up on that later on: https://github.com/dotnet/roslyn/issues/65394

**Comment by arkalyanms** [comment]
Merging based on QB response over email. This should be taken when 17.4.2 opens up.

**Comment by arkalyanms** [comment]
@Cosifne should we consider backporting the integration test fixes to 17.5 and enabling the check since we are taking servicing fixes? Might be worthwhile discussing in team meeting.

**Comment by sharwell** [comment]
@tmat do we know how this worked prior to 17.4?

**Comment by Youssef1313** [comment]
@sharwell Maybe because of the deletion of https://github.com/dotnet/roslyn/blob/da6f98ea75b91293a16ed2f75ac0d7bae767c4fd/src/VisualStudio/Core/Def/Options/LanguageSettingsPersister.cs in https://github.com/dotnet/roslyn/commit/51cc7c9a54246991828f789771f02157e981a37f

I'm not sure though how Visual Studio is interacting with these options, so @tmat can say for sure.

In the deleted class, I also see we were initializing settings for TS, F#, and Xaml. I don't know how these are supposed to work today with the current implementation.

**Comment by tmat** [comment]
Yup, we used a legacy VS interface to read the options.

**Comment by Youssef1313** [comment]
@tmat What about TS, F#, and Xaml? How do they work now?

**Comment by tmat** [comment]
When they call Roslyn to perform an operation that needs these options the operation reads them from global options, which works the same way as C#/VB.

**Pull request merged**
